File: desktop/lib/capture-store.js
import { appendFileSync, existsSync, mkdirSync, readdirSync, readFileSync, rmSync, writeFileSync } from 'node:fs';
import path from 'node:path';

export class DiskFullError extends Error {
  constructor() {
    super('No queda espacio para seguir grabando. El audio ya guardado se conserva.');
    this.code = 'ENOSPC';
  }
}

function emptyManifest(clientCaptureId, startedAt) {
  return {
    clientCaptureId,
    startedAt: startedAt || null,
    remoteConfirmed: false,
    audioStatus: 'partial',
    channelsComplete: false,
    channels: {},
  };
}

export class CaptureStore {
  constructor(root, io = {}) {
    this.root = root;
    this.io = {
      appendFile: io.appendFile || appendFileSync,
      mkdir: io.mkdir || mkdirSync,
      readFile: io.readFile || readFileSync,
      writeFile: io.writeFile || writeFileSync,
      readdir: io.readdir || readdirSync,
      rm: io.rm || rmSync,
      exists: io.exists || existsSync,
    };
    this.io.mkdir(root, { recursive: true });
  }

  begin(clientCaptureId, meta = {}) {
    const dir = this.#dir(clientCaptureId);
    this.io.mkdir(dir, { recursive: true });
    const file = this.#manifestPath(clientCaptureId);
    if (!this.io.exists(file)) {
      this.#write(emptyManifest(clientCaptureId, meta.startedAt));
    }
    return this.read(clientCaptureId);
  }

  append(clientCaptureId, channel, chunk) {
    this.begin(clientCaptureId);
    const manifest = this.read(clientCaptureId);
    const current = manifest.channels[channel] || { bytes: 0, absent: false };
    if (current.absent) {
      throw new Error(`El canal ${channel} está marcado como ausente`);
    }
    const audioPath = current.path || path.join(this.#dir(clientCaptureId), `${channel}.bin`);
    try {
      this.io.appendFile(audioPath, chunk);
    } catch (err) {
      if (err && (err.code === 'ENOSPC' || err instanceof DiskFullError)) throw new DiskFullError();
      throw err;
    }
    manifest.channels[channel] = {
      path: audioPath,
      bytes: (current.bytes || 0) + chunk.length,
      absent: false,
    };
    this.#refresh(manifest);
    this.#write(manifest);
    return manifest;
  }

  noteChannelAbsent(clientCaptureId, channel, reason) {
    this.begin(clientCaptureId);
    const manifest = this.read(clientCaptureId);
    const previous = manifest.channels[channel];
    manifest.channels[channel] = {
      ...(previous || {}),
      absent: true,
      reason: reason || 'absent',
    };
    this.#refresh(manifest);
    this.#write(manifest);
    return manifest;
  }

  read(clientCaptureId) {
    return JSON.parse(this.io.readFile(this.#manifestPath(clientCaptureId), 'utf8'));
  }

  pending() {
    return this.io
      .readdir(this.root)
      .map((name) => {
        try {
          return this.read(name);
        } catch {
          return null;
        }
      })
      .filter((row) => row && row.remoteConfirmed === false);
  }

  confirmRemote(clientCaptureId) {
    const manifest = this.read(clientCaptureId);
    manifest.remoteConfirmed = true;
    this.#write(manifest);
    return manifest;
  }

  discard(clientCaptureId) {
    const manifest = this.read(clientCaptureId);
    if (!manifest.remoteConfirmed) {
      throw new Error('No se borra la captura local antes de la confirmación remota');
    }
    this.io.rm(this.#dir(clientCaptureId), { recursive: true, force: true });
  }

  #refresh(manifest) {
    const mic = manifest.channels.mic;
    const system = manifest.channels.system;
    const micOk = Boolean(mic && !mic.absent && mic.bytes > 0);
    const systemOk = Boolean(system && !system.absent && system.bytes > 0);
    manifest.channelsComplete = micOk && systemOk;
    manifest.audioStatus = manifest.channelsComplete ? 'complete' : 'partial';
  }

  #dir(clientCaptureId) {
    return path.join(this.root, clientCaptureId);
  }

  #manifestPath(clientCaptureId) {
    return path.join(this.#dir(clientCaptureId), 'manifest.json');
  }

  #write(manifest) {
    this.io.mkdir(this.#dir(manifest.clientCaptureId), { recursive: true });
    this.io.writeFile(this.#manifestPath(manifest.clientCaptureId), JSON.stringify(manifest));
  }
}

File: desktop/macos/Sources/VocifyHostKit/CaptureStore.swift
import Foundation

public struct DiskFullError: Error, Equatable {
    public static let code = "ENOSPC"

    public init() {}

    public var localizedDescription: String {
        "No queda espacio para seguir grabando. El audio ya guardado se conserva."
    }
}

public final class CaptureStore {
    public let root: URL
    private let fileManager: FileManager
    private let customAppendFile: ((URL, Data) throws -> Void)?

    public init(
        root: URL,
        fileManager: FileManager = .default,
        appendFile: ((URL, Data) throws -> Void)? = nil
    ) {
        self.root = root.standardizedFileURL
        self.fileManager = fileManager
        self.customAppendFile = appendFile
        try? fileManager.createDirectory(at: self.root, withIntermediateDirectories: true)
    }

    public static func defaultCapturesRoot(fileManager: FileManager = .default) -> URL {
        let appSupport = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        return appSupport.appendingPathComponent("Vocify/captures", isDirectory: true)
    }

    @discardableResult
    public func begin(clientCaptureId: String, meta: [String: Any] = [:]) throws -> [String: Any] {
        let dir = captureDir(clientCaptureId)
        try fileManager.createDirectory(at: dir, withIntermediateDirectories: true)
        let manifestURL = manifestPath(clientCaptureId)
        if !fileManager.fileExists(atPath: manifestURL.path) {
            let startedAt = meta["startedAt"] as? String
            try write(emptyManifest(clientCaptureId: clientCaptureId, startedAt: startedAt), clientCaptureId: clientCaptureId)
        }
        return try read(clientCaptureId: clientCaptureId)
    }

    @discardableResult
    public func append(clientCaptureId: String, channel: String, chunk: Data) throws -> [String: Any] {
        _ = try begin(clientCaptureId: clientCaptureId)
        var manifest = try read(clientCaptureId: clientCaptureId)
        var channels = manifest["channels"] as? [String: Any] ?? [:]
        let current = channels[channel] as? [String: Any] ?? [:]
        if current["absent"] as? Bool == true {
            throw NSError(
                domain: "CaptureStore",
                code: 1,
                userInfo: [NSLocalizedDescriptionKey: "El canal \(channel) está marcado como ausente"]
            )
        }

        let dir = captureDir(clientCaptureId)
        let audioPath = (current["path"] as? String).map { URL(fileURLWithPath: $0) }
            ?? dir.appendingPathComponent("\(channel).bin")

        do {
            try appendData(to: audioPath, chunk: chunk)
        } catch let error as DiskFullError {
            throw error
        } catch {
            if Self.isDiskFull(error) {
                throw DiskFullError()
            }
            throw error
        }

        let previousBytes = current["bytes"] as? Int ?? 0
        channels[channel] = [
            "path": audioPath.path,
            "bytes": previousBytes + chunk.count,
            "absent": false,
        ] as [String: Any]
        manifest["channels"] = channels
        refresh(&manifest)
        try write(manifest, clientCaptureId: clientCaptureId)
        return manifest
    }

    @discardableResult
    public func noteChannelAbsent(clientCaptureId: String, channel: String, reason: String?) throws -> [String: Any] {
        _ = try begin(clientCaptureId: clientCaptureId)
        var manifest = try read(clientCaptureId: clientCaptureId)
        var channels = manifest["channels"] as? [String: Any] ?? [:]
        let previous = channels[channel] as? [String: Any] ?? [:]
        var next = previous
        next["absent"] = true
        next["reason"] = reason ?? "absent"
        channels[channel] = next
        manifest["channels"] = channels
        refresh(&manifest)
        try write(manifest, clientCaptureId: clientCaptureId)
        return manifest
    }

    public func read(clientCaptureId: String) throws -> [String: Any] {
        let data = try Data(contentsOf: manifestPath(clientCaptureId))
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw NSError(domain: "CaptureStore", code: 2, userInfo: [NSLocalizedDescriptionKey: "Invalid manifest"])
        }
        return json
    }

    public func pending() -> [[String: Any]] {
        guard let names = try? fileManager.contentsOfDirectory(atPath: root.path) else { return [] }
        return names.compactMap { name -> [String: Any]? in
            do {
                let manifest = try read(clientCaptureId: name)
                if manifest["remoteConfirmed"] as? Bool == false {
                    return manifest
                }
                return nil
            } catch {
                return nil
            }
        }
    }

    @discardableResult
    public func confirmRemote(clientCaptureId: String) throws -> [String: Any] {
        var manifest = try read(clientCaptureId: clientCaptureId)
        manifest["remoteConfirmed"] = true
        try write(manifest, clientCaptureId: clientCaptureId)
        return manifest
    }

    public func discard(clientCaptureId: String) throws {
        let manifest = try read(clientCaptureId: clientCaptureId)
        if manifest["remoteConfirmed"] as? Bool != true {
            throw NSError(
                domain: "CaptureStore",
                code: 3,
                userInfo: [NSLocalizedDescriptionKey: "No se borra la captura local antes de la confirmación remota"]
            )
        }
        try fileManager.removeItem(at: captureDir(clientCaptureId))
    }

    private func emptyManifest(clientCaptureId: String, startedAt: String?) -> [String: Any] {
        [
            "clientCaptureId": clientCaptureId,
            "startedAt": startedAt as Any? ?? NSNull(),
            "remoteConfirmed": false,
            "audioStatus": "partial",
            "channelsComplete": false,
            "channels": [String: Any](),
        ]
    }

    private func refresh(_ manifest: inout [String: Any]) {
        let channels = manifest["channels"] as? [String: Any] ?? [:]
        let mic = channels["mic"] as? [String: Any]
        let system = channels["system"] as? [String: Any]
        let micOk = channelHasAudio(mic)
        let systemOk = channelHasAudio(system)
        manifest["channelsComplete"] = micOk && systemOk
        manifest["audioStatus"] = (micOk && systemOk) ? "complete" : "partial"
    }

    private func channelHasAudio(_ channel: [String: Any]?) -> Bool {
        guard let channel else { return false }
        if channel["absent"] as? Bool == true { return false }
        return (channel["bytes"] as? Int ?? 0) > 0
    }

    private func captureDir(_ clientCaptureId: String) -> URL {
        root.appendingPathComponent(clientCaptureId, isDirectory: true)
    }

    private func manifestPath(_ clientCaptureId: String) -> URL {
        captureDir(clientCaptureId).appendingPathComponent("manifest.json")
    }

    private func write(_ manifest: [String: Any], clientCaptureId: String) throws {
        try fileManager.createDirectory(at: captureDir(clientCaptureId), withIntermediateDirectories: true)
        let data = try JSONSerialization.data(withJSONObject: manifest)
        try data.write(to: manifestPath(clientCaptureId), options: .atomic)
    }

    private func appendData(to url: URL, chunk: Data) throws {
        if let customAppendFile {
            try customAppendFile(url, chunk)
            return
        }
        if fileManager.fileExists(atPath: url.path) {
            let handle = try FileHandle(forWritingTo: url)
            defer { try? handle.close() }
            try handle.seekToEnd()
            try handle.write(contentsOf: chunk)
        } else {
            try chunk.write(to: url, options: .atomic)
        }
    }

    private static func isDiskFull(_ error: Error) -> Bool {
        if error is DiskFullError { return true }
        let ns = error as NSError
        if ns.domain == NSPOSIXErrorDomain, ns.code == Int(POSIXError.ENOSPC.rawValue) { return true }
        if ns.domain == NSCocoaErrorDomain, ns.code == NSFileWriteOutOfSpaceError { return true }
        if let underlying = ns.userInfo[NSUnderlyingErrorKey] as? NSError {
            return isDiskFull(underlying)
        }
        return false
    }
}
